api/server/runner_httptrigger.go

```diff
@@ -112,15 +112,23 @@ func (trw *triggerResponseWriter) WriteHeader(statusCode int) {
 		realHeaders[k] = vs
 	}
 
-	// XXX(reed): simplify / add tests for these behaviors...
-	gatewayStatus := 200
-	if statusCode >= 400 {
-		gatewayStatus = 502
-	} else if fnStatus > 0 {
-		gatewayStatus = fnStatus
+	trw.inner.WriteHeader(determineFinalStatus(statusCode, fnStatus))
+}
+
+func determineFinalStatus(serviceStatus, userFnStatus int) int {
+	finalStatus := 200
+
+	if userFnStatus > 0 {
+		if userFnStatus >= 500 {
+			finalStatus = 502
```

Comment:
Hmm - not super hot on this  

My 2c : 
  - if the function failed to execute correctly we should always return a 502 irrespective of the user status (possibly with exception of 400 which we might honour) 
 -  if the user gave us a status and the function wasn't in error we should honour it 

if we're depending on HTTP status for much downstream then we should fix this.

Reply:
So you'd like to see

if userFnStatus >= 400 {
  finalStatus = 502
}

I dunno, I guess i'm just not that bothered about 4xx, let them send them if they want.

Reply:
hmm actually I take it back I think the above is OK -

Reply:
I think one of the things i want to be able to implement in triggers is REST apis - so any <500 codes should be honoured  - agree on angling towards coalescing all >=500 codes to 502 though

Reply:
I think one of the things i want to be able to implement in triggers is REST apis - so any <500 codes should be honoured  - agree on angling towards coalescing all >=500 codes to 502 though

Reply:
that would maybe be okay as a default policy on an API gateway that a user could modify but for users that want to run a website off of fn http triggers with their own 500 codes that their services use to interact with one another, this seems like way too low of a level to be rewriting status codes, especially on the basis that the only reason we're staring at these things is because we don't want our stats blowing up when users return their own codes - this seems a selfish and lazy solution to take. say, they retry on 503, now they suddenly get every 501 and 503 as a 502 and they have to encode the actual code in their error message (what are status codes for anyway?) - this is _at best_ frustrating for users to realize that they wish they had found this so that they didn't waste so much time trying to integrate to realize they have to go change all their error messages to migrate over to functions and it's going to be a larger migration than they originally thought. without any way for users to return their own codes in some capacity (whether it's via an api gateway or just straight from their function) I think this solution of rewriting >500 to 502 is not justifiable - and alternative solutions exist to solve the problem we are having currently.